config/src/main/java/org/springframework/security/config/annotation/web/configurers/oauth2/server/resource/OAuth2ResourceServerConfigurer.java

```diff
@@ -154,12 +166,26 @@ public void configure(H http) throws Exception {
 
 		private JwtConfigurer() {}
 
+		public OAuth2ResourceServerConfigurer<H> decoder(JwtDecoder decoder) {
+			this.decoder = decoder;
+			return OAuth2ResourceServerConfigurer.this;
+		}
+
 		public OAuth2ResourceServerConfigurer<H> jwkSetUri(String uri) {
 			this.decoder = new NimbusJwtDecoderJwkSupport(uri);
 			return OAuth2ResourceServerConfigurer.this;
 		}
 
-		private JwtDecoder getJwtDecoder() {
+		private JwtDecoder getJwtDecoder(H http) {
+			if ( this.decoder == null ) {
+				ApplicationContext context = http.getSharedObject(ApplicationContext.class);
+				Map<String, JwtDecoder> decoders =
```

Comment:
It is surprising if there are multiple `JwtDecoder` beans that we just use the first one. This can make troubleshooting issues pretty tricky. I think we should just rely on `context.getBean(JwtDecoder.class)` if the user did not explicitly configure the `JwtDecoder` as this will give a reasonable error message that we need a `JwtDecoder`.

Reply:
We should add a tests for the cases where

* there are two `JwtDecoder` beans and it is explicitly configured via the DSL (expected to work with the configured value)
* there are two `JwtDecoder` beans and it is not explicitly configured via the DSL (should produce an error since it is ambiguous which bean to use)

Reply:
That is certainly simpler, and yes I'd prefer leaving that resolution logic to someone else.

I'm curious because I see this pattern of calling BeanFactoryUtils in other configurers (that is, call this same method, and take the first value from the map), and I wonder if there is good reason for using this pattern elsewhere?

Reply:
It depends on if it needs to be lazy or not. For example AuthenticationConfiguration uses the `BeanFactoryUtils` to create a lazy implementation of an object because looking the bean up at that point in time is likely to cause a cycle. I noticed that OAuth2LoginConfigurer is using BeanFactoryUtils...this appears to be a mistake from what I can tell. Can you create a ticket for it?

Reply:
@jzheaux Have you seen `OAuth2ClientConfigurerUtils`? It handles the case where there is more than 1 `@Bean` registered and throws a descriptive exception. The same logic would apply.

@rwinch Can you expand on the issue using `BeanFactoryUtils.beansOfTypeIncludingAncestors` in `OAuth2ClientConfigurerUtils`?

Reply:
@jzheaux I'm not sure what advantage this brings to just `context.getBean(clazz)`. It also throws an informative error message.

Reply:
fyi - https://github.com/spring-projects/spring-security/issues/5538